src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNVectorScriptDocValues.java

```diff
@@ -0,0 +1,72 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.knn.index;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+
+// This class is thread safe, since docExists is synchronized at an instance level
+public final class KNNVectorScriptDocValues extends ScriptDocValues<float[]> {
+
+    private final BinaryDocValues binaryDocValues;
+    private boolean docExists;
+
+    public KNNVectorScriptDocValues(BinaryDocValues binaryDocValues) {
+        this.binaryDocValues = binaryDocValues;
+    }
+
+    @Override
+    public void setNextDocId(int docId) throws IOException {
+        synchronized (this) {
```

Comment:
I added thread safe to prevent any race condition based on feedback, since docExists setting is happening in one place and reading is happening on another place. If there is not threat with race condition then definitely we can remove this. At the end, it comes to do we need to make this instance thread safe or not? What do you think?

Reply:
Good point. Do you have pointers to any other base classes extending ``` ScriptDocValues<>``` doing synchronization? My understanding is each search request should have its own instance and documents are iterated sequentially so synchronization should not be required. If you see any base class doing this we can definitely consider,

Reply:
Removed synchronization.